File: src/lib/api/idle.ts
import { invoke } from "@tauri-apps/api/core";
import type { IdleSnapshot, WeatherLocationCandidate } from "./types";

let snapshot: { key: string; at: number; value: IdleSnapshot } | undefined;
let pending: { key: string; request: Promise<IdleSnapshot> } | undefined;
export const idleApi = {
  getSnapshot: (key = "") => {
    if (snapshot?.key === key && Date.now() - snapshot.at < 30_000) return Promise.resolve(snapshot.value);
    if (pending?.key === key) return pending.request;
    const request = invoke<IdleSnapshot>("get_idle_snapshot").then(value => {
      if (pending?.request === request) snapshot = {key, at: Date.now(), value};
      return value;
    }).finally(() => { if (pending?.request === request) pending = undefined; });
    pending = {key, request};
    return request;
  },
  searchLocations: (query: string, language = "zh") => invoke<WeatherLocationCandidate[]>("search_weather_locations", { query, language }),
};

File: src-tauri/src/commands/idle.rs
use crate::error::AppResult;
use crate::models::{IdleSnapshot, WeatherLocationCandidate};
use tauri::AppHandle;

#[tauri::command]
pub async fn get_idle_snapshot(app: AppHandle) -> AppResult<IdleSnapshot> {
    crate::services::get_idle_snapshot(&app).await
}

#[tauri::command]
pub async fn search_weather_locations(
    query: String,
    language: String,
) -> AppResult<Vec<WeatherLocationCandidate>> {
    crate::services::search_weather_locations(&query, &language).await
}
